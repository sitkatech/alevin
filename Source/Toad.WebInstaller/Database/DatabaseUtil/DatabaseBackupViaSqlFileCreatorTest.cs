using System.Data.SqlClient;
using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;

namespace Toad.WebInstaller.Database.DatabaseUtil
{
    [TestFixture]
    public class DatabaseBackupViaSqlFileCreatorTest
    {
        private const string DatabaseServer = "(local)";
        private const string DefaultStartDatabase = "master";
        private readonly string _testDatabaseName;

        public DatabaseBackupViaSqlFileCreatorTest()
        {
            _testDatabaseName = GetType().Name;
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void CanScriptOutDatabase()
        {
            // Arrange
            // -------
            using (var initialConn = new SqlConnection(string.Format("Server={0};Database={1};Trusted_Connection=True;pooling=false", DatabaseServer, DefaultStartDatabase)))
            {
                initialConn.Open();
                CleanupDatabase(initialConn);
                SetupTestDatabase(initialConn);
                initialConn.Close();
            }

            try
            {
                var testSubject = new DatabaseBackupViaSqlFileCreator(DatabaseServer, _testDatabaseName);

                // Act
                // ---
                string sqlBackup = testSubject.CreateSqlBackup();

                // Assert
                // ------
                Approvals.Verify(sqlBackup);
            }
            finally
            {
                using (var secondConn = new SqlConnection(string.Format("Server={0};Database={1};Trusted_Connection=True;pooling=false", DatabaseServer, DefaultStartDatabase)))
                {
                    secondConn.Open();
                    // Clean up
                    // ---------
                    CleanupDatabase(secondConn);
                    secondConn.Close();
                }
            }
        }

        private void SetupTestDatabase(SqlConnection initialConn)
        {
            string createTestDatabaseSql = string.Format("create database {0}", _testDatabaseName);
            using (var command = new SqlCommand(createTestDatabaseSql, initialConn))
            {
                command.ExecuteNonQuery();
            }

            using (var testDatabaseConn = new SqlConnection(string.Format("Server={0};Database={1};Trusted_Connection=True;pooling=false", DatabaseServer, _testDatabaseName)))
            {
                testDatabaseConn.Open();
                const string setupTablesSql = @"
create table tblB1
(
    ColB1Int int not null,
    ColC1Int int not null,
    ColB1String varchar(100) not null,
    ColB1Binary varbinary(100) not null,
    constraint PK_tblB1_ColB1Int primary key (ColB1Int),
);

create table tblC1
(
    ColC1Int int not null identity(1,1),
    ColC1String varchar(100),
    ColC1Binary varbinary(100),
    constraint PK_tblC1_ColC1Int primary key (ColC1Int)
);

CREATE TABLE [dbo].[sysdiagrams](
	[name] [sysname] NOT NULL,
	[principal_id] [int] NOT NULL,
	[diagram_id] [int] IDENTITY(1,1) NOT NULL,
	[version] [int] NULL,
	[definition] [varbinary](max) NULL,
PRIMARY KEY CLUSTERED 
(
	[diagram_id] ASC
)WITH (PAD_INDEX  = OFF, STATISTICS_NORECOMPUTE  = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS  = ON, ALLOW_PAGE_LOCKS  = ON),
    CONSTRAINT [UK_principal_name] UNIQUE NONCLUSTERED 
(
	[principal_id] ASC,
	[name] ASC
)WITH (PAD_INDEX  = OFF, STATISTICS_NORECOMPUTE  = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS  = ON, ALLOW_PAGE_LOCKS  = ON)
) 
EXEC sys.sp_addextendedproperty @name=N'microsoft_database_tools_support', @value=1 , @level0type=N'SCHEMA',@level0name=N'dbo', @level1type=N'TABLE',@level1name=N'sysdiagrams'

alter table tblB1 add constraint FK_tblB1_ColC1Int_tblC1_ColC1Int foreign key (ColC1Int) references tblC1(ColC1Int);


set identity_insert tblC1 on
insert into tblC1(ColC1Int, ColC1String, ColC1Binary) values (1, 'Hi there ╣ unicode', 0x012345);
insert into tblC1(ColC1Int, ColC1String, ColC1Binary) values (2, 'Hi there Some More', 0x0123454);
set identity_insert tblC1 off

insert into tblB1 (ColB1Int, ColC1Int, ColB1String, ColB1Binary) values (1001, 1, 'Hi there', 0x012345);
";

                using (var command = new SqlCommand(setupTablesSql, testDatabaseConn))
                {
                    command.ExecuteNonQuery();
                }
                testDatabaseConn.Close();
            }




        }

        private void CleanupDatabase(SqlConnection conn)
        {
            string clearupDatabaseSql = string.Format(@"
                if exists (select * from sys.databases as sd where sd.name = '{0}')
                begin
                    exec('use {0}
                    -- Erland Sommarskog suggests this before the drop; otherwise we get errors about the database still being in use.
                    ALTER DATABASE {0} SET SINGLE_USER WITH ROLLBACK IMMEDIATE
                    USE master

                    drop database {0}')
                end
                ",  _testDatabaseName);
            
            using (var command = new SqlCommand(clearupDatabaseSql, conn))
            {
                command.ExecuteNonQuery();
            }
        }

    }
}