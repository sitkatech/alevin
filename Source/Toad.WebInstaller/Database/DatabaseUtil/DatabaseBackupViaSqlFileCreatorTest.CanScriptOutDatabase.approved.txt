SET ANSI_NULLS ON
SET QUOTED_IDENTIFIER ON
CREATE TABLE [dbo].[sysdiagrams](
	[name] [sysname] COLLATE SQL_Latin1_General_CP1_CI_AS NOT NULL,
	[principal_id] [int] NOT NULL,
	[diagram_id] [int] IDENTITY(1,1) NOT NULL,
	[version] [int] NULL,
	[definition] [varbinary](max) NULL,
PRIMARY KEY CLUSTERED 
(
	[diagram_id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY],
 CONSTRAINT [UK_principal_name] UNIQUE NONCLUSTERED 
(
	[principal_id] ASC,
	[name] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]

EXEC sys.sp_addextendedproperty @name=N'microsoft_database_tools_support', @value=1 , @level0type=N'SCHEMA',@level0name=N'dbo', @level1type=N'TABLE',@level1name=N'sysdiagrams'
SET ANSI_NULLS ON
SET QUOTED_IDENTIFIER ON
CREATE TABLE [dbo].[tblC1](
	[ColC1Int] [int] IDENTITY(1,1) NOT NULL,
	[ColC1String] [varchar](100) COLLATE SQL_Latin1_General_CP1_CI_AS NULL,
	[ColC1Binary] [varbinary](100) NULL,
 CONSTRAINT [PK_tblC1_ColC1Int] PRIMARY KEY CLUSTERED 
(
	[ColC1Int] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY]

SET ANSI_NULLS ON
SET QUOTED_IDENTIFIER ON
CREATE TABLE [dbo].[tblB1](
	[ColB1Int] [int] NOT NULL,
	[ColC1Int] [int] NOT NULL,
	[ColB1String] [varchar](100) COLLATE SQL_Latin1_General_CP1_CI_AS NOT NULL,
	[ColB1Binary] [varbinary](100) NOT NULL,
 CONSTRAINT [PK_tblB1_ColB1Int] PRIMARY KEY CLUSTERED 
(
	[ColB1Int] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY]

ALTER TABLE [dbo].[tblB1]  WITH CHECK ADD  CONSTRAINT [FK_tblB1_ColC1Int_tblC1_ColC1Int] FOREIGN KEY([ColC1Int])
REFERENCES [dbo].[tblC1] ([ColC1Int])
ALTER TABLE [dbo].[tblB1] CHECK CONSTRAINT [FK_tblB1_ColC1Int_tblC1_ColC1Int]
GO
SET IDENTITY_INSERT [dbo].[tblC1] ON 

INSERT [dbo].[tblC1] ([ColC1Int], [ColC1String], [ColC1Binary]) VALUES (1, N'Hi there ¦ unicode', 0x012345)
INSERT [dbo].[tblC1] ([ColC1Int], [ColC1String], [ColC1Binary]) VALUES (2, N'Hi there Some More', 0x00123454)
SET IDENTITY_INSERT [dbo].[tblC1] OFF
INSERT [dbo].[tblB1] ([ColB1Int], [ColC1Int], [ColB1String], [ColB1Binary]) VALUES (1001, 1, N'Hi there', 0x012345)